File: app/src/main/java/uk/co/victoriajanedavis/chatapp/domain/interactors/LoginUser.java
package uk.co.victoriajanedavis.chatapp.domain.interactors;

import javax.inject.Inject;

import io.reactivex.Single;
import io.reactivex.annotations.NonNull;
import uk.co.victoriajanedavis.chatapp.data.repositories.TokenRepository;
import uk.co.victoriajanedavis.chatapp.domain.entities.TokenEntity;
import uk.co.victoriajanedavis.chatapp.domain.interactors.ReactiveInteractor.SendInteractor;

public class LoginUser implements SendInteractor<LoginUser.LoginParams, TokenEntity> {

    @NonNull private final TokenRepository backendTokenRepo;


    @Inject
    public LoginUser(@NonNull final TokenRepository backendTokenRepo) {
        this.backendTokenRepo = backendTokenRepo;
    }

    @NonNull
    @Override
    public Single<TokenEntity> getSingle(@NonNull LoginParams loginParams) {
        return backendTokenRepo.fetchTokenViaLogin(loginParams.username, loginParams.password)
                .andThen(backendTokenRepo.requestTokenSingle());
    }

    public static final class LoginParams {
        @NonNull private final String username;
        @NonNull private final String password;

        public LoginParams(String username, String password) {
            this.username = username;
            this.password = password;
        }
    }
}

File: app/src/main/java/uk/co/victoriajanedavis/chatapp/domain/interactors/LoginUser.kt
package uk.co.victoriajanedavis.chatapp.domain.interactors

import io.reactivex.Completable
import javax.inject.Inject

import uk.co.victoriajanedavis.chatapp.data.repositories.FirebaseTokenRepository
import uk.co.victoriajanedavis.chatapp.data.repositories.TokenRepository
import uk.co.victoriajanedavis.chatapp.domain.interactors.ReactiveInteractor.ActionInteractor

class LoginUser @Inject constructor(
    private val backendTokenRepo: TokenRepository,
    private val firebaseTokenRepo: FirebaseTokenRepository
) : ActionInteractor<LoginUser.LoginParams> {

    // If POST of the Firebase registration token fails then whole login process fails
    override fun getActionCompletable(params: LoginParams): Completable {
        return backendTokenRepo.fetchTokenViaLoginAndStore(params.username, params.password)
            .andThen(firebaseTokenRepo.postTokenToBackend()
                .onErrorResumeNext(::deleteTokenFromStorageThenError)
            )
    }

    private fun deleteTokenFromStorageThenError(e: Throwable): Completable {
        return backendTokenRepo.deleteTokenFromStorage()
            .andThen { Completable.error(Exception("Login failed: ${e.message}")) }
    }

    class LoginParams(
        val username: String,
        val password: String
    )
}
